lib/naxum/src/serve.rs

```diff
@@ -176,32 +177,30 @@ where
                         tracker.close();
                         break;
                     }
-                    (msg, permit) = next_message(&mut stream, semaphore.clone(), failed_count) => {
+                    (msg, permit) = next_message(&mut stream, semaphore.clone()) => {
                         match msg {
                             Some(Ok(msg)) => {
-                                failed_count = 0;
-                                metric!(counter.naxum.failed_count = 0);
-
+                                metric!(counter.naxum.next_message.failed = 0);
+                                metric!(counter.naxum.next_message.none = 0);
                                 (msg, permit)
                             },
                             Some(Err(err)) => {
-                                // TODO(fnichol): this level might need to be `trace!()`, just
-                                // unclear at the moment
-                                warn!(si.error.message = ?err, "failed to read next message from stream");
-                                failed_count += 1;
+                                error!(si.error.message = ?err, "failed to read next message from stream");
+                                metric!(counter.naxum.next_message.failed = 1);
                                 continue;
                             },
                             None => {
-                                info!("stream is closed, breaking out of loop");
-                                tracker.close();
-                                break;
+                                warn!(sleep_duration_ms = ?NEXT_NONE_MESSAGE_SLEEP_DURATION, "next message is 'None' on provided stream: going to sleep");
+                                metric!(counter.naxum.next_message.none = 1);
+                                std::thread::sleep(NEXT_NONE_MESSAGE_SLEEP_DURATION);
+                                continue;
```

Comment:
I'm not 100% sure on this, I need to check the rest of the service flow. Reason being, this isn't what an HTTP server would do and while I know that's not what this is, we have no way to know whether this incoming `futures::Stream` would ever yield another `Some(_)` value, usually the streams are fused and thus permanently closed.

Reply:
We spoke the folks managing the client and were told reconnection could cause a `Ready(None)` to be followed by an eventual `Ready(Some(_))`.

Despite that, you and I just caught up and believe that we would rather follow convention and shutdown here, otherwise many naxum apps would be following anti-patterns in shutdown procedures.